Translation for Updating Objects in State page (#886)
reactjs/ru.react.dev: src/content/learn/updating-objects-in-state.md

---
title: Обновление объектов в состоянии
---

<Intro>

Состояние может содержать в себе любые JavaScript значения, включая объекты. Значения объектов, которые находятся в состоянии, нельзя изменять напрямую. Вместо этого, если вы хотите обновить состояние, вам необходимо создать новый объект или копию текущего объекта, а затем установить состоянию этот объект.

</Intro>

<YouWillLearn>

- Как правильно в React обновлять объект в состоянии
- Как обновить вложенный объект без мутации
- Что такое иммутабельность, как её не нарушить
- Как сделать копирование объектов менее повторяющимся и сложным с Immer

</YouWillLearn>

## Что такое мутация? {/*whats-a-mutation*/}

Вы можете хранить любое JavaScript значение в состоянии.

```js
const [x, setX] = useState(0);
```

> До сих пор вы работали с числами, строками и логическими значениями. Эти виды значений в JavaScript являются "иммутабельными" -- это значит, что они не изменяются и доступны "только для чтения". Вы можете запустить повторный рендер для _замены_ значения:

Author: а если так поступить: "являются иммутабельными (далее — неизменяемыми)"
и просто работать с "неизменяемый" дальше по тексту, мы так два варианта определения покажем сразу

> До сих пор вы работали с числами, строками и логическими значениями. Эти виды значений в JavaScript являются "иммутабельными" -- это значит, что они не изменяются и доступны "только для чтения". Вы можете запустить повторный рендер для _замены_ значения:

Reviewer: мне кажется, это хороший подход

> До сих пор вы работали с числами, строками и логическими значениями. Эти виды значений в JavaScript являются "иммутабельными" -- это значит, что они не изменяются и доступны "только для чтения". Вы можете запустить повторный рендер для _замены_ значения:

Reviewer: новая вводная по поводу этого комментария - в соглашении этого репозитория всё же есть договорённость переводить "immutable" как "иммутабельный". Я просмотрел, потому что они приведены чуть ниже на букву m. Так что извините меня за переполох, давайте придерживаться этого варианта по всему тексту. Хорошо, если еще не правили :)
Впрочем, по-прежнему можно добавить "неизменяемый" в скобках при первом использовании. На ваше усмотрение.

> До сих пор вы работали с числами, строками и логическими значениями. Эти виды значений в JavaScript являются "иммутабельными" -- это значит, что они не изменяются и доступны "только для чтения". Вы можете запустить повторный рендер для _замены_ значения:

Author: спасибо за информацию! все в порядке, в ближайшее время отправлю новый вариант работы :)

});
```

Мутация становится проблемой, когда вы изменяете *существующий* объект, который уже в состоянии. Мутировать объект, который вы только что создали, допустимо, так как *никакой код не ссылается на этот объект*. От изменения такого объекта ничего не зависит, он ничего не может случайно сломать. Это и есть "локальная мутация". Вы даже можете проводить локальную мутацию объекта [пока идёт рендер.](/learn/keeping-components-pure#local-mutation-your-components-little-secret) Это удобно и совершенно нормально!

</DeepDive>  

> ## Копирование объектов с использованием спред-синтаксиса {/*copying-objects-with-the-spread-syntax*/}

Reviewer: ```suggestion
## Копирование объектов с использованием оператора расширения {/*copying-objects-with-the-spread-syntax*/}
```
Здесь и ниже в переводе я бы рекомендовал использовать [оператор расширения ](https://learn.javascript.ru/rest-parameters-spread-operator). Или синтаксис расширения.

> В предыдущем примере, объект `position` всегда создаётся заново, исходя из текущей позиции курсора. На практике чаще всего вы захотите включать уже *существующие* данные в новый объект, который вы создаёте. Например, вы можете захотеть обновить *только одно* поле в форме, остальные значения полей сохранить без изменений.

Reviewer: ```suggestion
В предыдущем примере объект `position` всегда создаётся заново, исходя из текущей позиции курсора. На практике чаще всего вы захотите включать уже *существующие* данные в новый объект, который вы создаёте. Например, вы можете захотеть обновить *только одно* поле в форме, остальные значения полей сохранить без изменений.
```
В русском не обособляется

Эти поля ввода не работают, потому что обработчики `onChange` изменяют состояние:

> 

Reviewer: Вы могли бы перевести примеры и в этом переводе?

</Sandpack>

Например, эта строка изменяет состояние из прошлого рендера:

```js
person.firstName = e.target.value;
```

Надёжный способ получить желаемое поведение -- создать новый объект и передать его в `setPerson`. Но здесь вы также захотите **скопировать в него существующие данные**, потому что изменилось только одно из полей:

```js
setPerson({
  firstName: e.target.value, // новое значение имени из поля ввода
  lastName: person.lastName,
  email: person.email
});
```

> Вы можете использовать `...` [спред-оператор,](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals) чтобы вам не нужно было копировать каждое свойство отдельно.

Reviewer: ```suggestion
Вы можете использовать `...` [оператор расширения,](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_%D0%B2_%D0%BB%D0%B8%D1%82%D0%B5%D1%80%D0%B0%D0%BB%D0%B0%D1%85_%D0%BE%D0%B1%D1%8A%D0%B5%D0%BA%D1%82%D0%B0) чтобы вам не нужно было копировать каждое свойство отдельно.
```
Обновил ссылку на русскоязычный вариант статьи в mdn.

};
```

Если бы вы изменили `obj3.artwork.city`, это повлияло бы как на `obj2.artwork.city`, так и на `obj1.city`. Это связано с тем, что `obj3.artwork`, `obj2.artwork` и `obj1` являются одним и тем же объектом. Это трудно осознать, когда вы думаете об объектах как о «вложенных». Вместо этого они представляют собой отдельные объекты, «указывающие» друг на друга.

</DeepDive>  

### Напишем лаконичное обновление с помощью Immer {/*write-concise-update-logic-with-immer*/}

> Если ваш объект имеет несколько уровней вложенности, вы можете подумать об [упрощении структуры объекта.](/learn/choosing-the-state-structure#avoid-deeply-nested-state) Однако, если вы не хотите изменять структуру объекта, вы можете ссылаться на вложенные объекты. [Immer](https://github.com/immerjs/use-immer) -- это популярная библиотека, которая позволяет вам использовать удобный, но мутирующий синтаксис, кроме того, она ответственна за создание копий объекта. С Immer ваш код выглядит так, будто вы "нарушили правило" и мутировали объект:

Reviewer: ```suggestion
Если ваш объект имеет несколько уровней вложенности, вы можете подумать об [упрощении структуры объекта.](/learn/choosing-the-state-structure#avoid-deeply-nested-state) Однако, если вы не хотите изменять структуру объекта, вы можете ссылаться на вложенные объекты. [Immer](https://github.com/immerjs/use-immer) -- это популярная библиотека, которая позволяет вам использовать удобный, но мутирующий синтаксис. Кроме того, она ответственна за создание копий объекта. С Immer ваш код выглядит так, будто вы "нарушили правило" и мутировали объект:
```
может уже лучше разбить на два предложения?

```js
updatePerson(draft => {
  draft.artwork.city = 'Lagos';
});
```

В отличие от обычной мутации, Immer не перезаписывает предыдущее состояние напрямую!

<DeepDive>

#### Как работает Immer? {/*how-does-immer-work*/}

> `draft`, предоставленный Immer, является особенным типом объекта, который называется [прокси](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), он "записывает" то, что вы с ним делаете. Вот почему вы можете мутировать его столько раз, сколько захотите! Под капотом Immer выясняет, какие части `draft` были изменены, и создаёт новый объект с наличием этих изменений.

Reviewer: ```suggestion
`Черновик`, предоставленный Immer, является особенным типом объекта, который называется [прокси](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), он "записывает" то, что вы с ним делаете. Вот почему вы можете мутировать его столько раз, сколько захотите! Под капотом Immer выясняет, какие части `черновика` были изменены, и создаёт новый объект с наличием этих изменений.
```
На мой взгляд, здесь можно перевести. По поиску "Immer черновик" находятся статьи на хабре и хекслете

* **Более простая реализация:** поскольку React не полагается на мутацию, ему не нужно делать ничего особенного с вашими объектами: не нужно захватывать их свойства, всегда оборачивать их в прокси или выполнять другую работу при инициализации, как это делают многие «реактивные» решения. По этой же причине React позволяет помещать в состояние любой объект, независимо от его размера, без дополнительных проблем с производительностью или корректностью.

На практике вы часто можете "уйти" от мутации состояния в React, но мы настоятельно рекомендуем вам не делать этого, чтобы вы могли использовать новые функции React, разработанные с учётом этого подхода.

</DeepDive>

<Recap>

* Рассматривайте все состояния в React как иммутабельные (неизменяемые).
* Когда вы храните объекты в состоянии, их прямое изменение не вызовет повторный рендер и изменит состояние в предыдущих "снимках" рендера.

> * Вместо этого, создайте *новую* версию объекта и активируйте повторный рендер, установив для него состояние.

Reviewer: ```suggestion
* Вместо этого создайте *новую* версию объекта и активируйте повторный рендер, установив для него состояние.
```
Кажется, тут тоже не нужно обособление
